spatialmath/mesh.go

```diff
@@ -29,10 +40,190 @@ func (m *Mesh) Triangles() []*Triangle {
+	for _, tri := range m.triangles {
+		points = append(points, tri.Points()...)
+	}
+
+	for _, pt := range points {
+		collides, err := NewPoint(pt, "").CollidesWith(g, defaultCollisionBufferMM)
+		if err != nil {
+			return false, err
+		}
+		if !collides {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+// DistanceFrom returns the minimum distance between this mesh and another geometry.
+func (m *Mesh) DistanceFrom(g Geometry) (float64, error) {
+	if other, ok := g.(*box); ok {
+		return m.distanceFromMesh(other.toMesh()), nil
+	}
+	if other, ok := g.(*capsule); ok {
+		return capsuleVsMeshDistance(other, m), nil
+	}
+	if other, ok := g.(*point); ok {
+		return m.distanceFromSphere(other.position, 0), nil
+	}
+	if other, ok := g.(*sphere); ok {
+		return m.distanceFromSphere(other.pose.Point(), other.radius), nil
+	}
+	if other, ok := g.(*Mesh); ok {
+		return m.distanceFromMesh(other), nil
+	}
+	return math.Inf(-1), newCollisionTypeUnsupportedError(m, g)
+}
+
+func (m *Mesh) distanceFromSphere(pt r3.Vector, radius float64) float64 {
+	minDist := math.Inf(1)
+	for _, tri := range m.triangles {
+		closestPt, _ := ClosestTriangleInsidePoint(tri, pt)
+		dist := closestPt.Sub(pt).Norm() - radius
+		if dist < minDist {
+			minDist = dist
+		}
+	}
+	return minDist
+}
+
+func (m *Mesh) collidesWithSphere(pt r3.Vector, radius, buffer float64) bool {
+	for _, tri := range m.triangles {
+		closestPt, inside := ClosestTriangleInsidePoint(tri, pt)
+		if inside && closestPt.Sub(pt).Norm() <= radius+buffer {
+			return true
+		}
+	}
+	return false
+}
+
+// collidesWithMesh checks if this mesh collides with another mesh
+// TODO: This function is *begging* for GPU acceleration.
+func (m *Mesh) collidesWithMesh(other *Mesh, collisionBufferMM float64) bool {
+	// Check if any triangles from either mesh collide
+	for _, tri1 := range m.triangles {
+		for _, tri2 := range other.triangles {
+			for _, pt := range tri1.Points() {
+				closestPt, inside := ClosestTriangleInsidePoint(tri2, pt)
+				if inside && closestPt.Sub(pt).Norm() <= collisionBufferMM {
+					return true
+				}
+			}
+			for _, pt := range tri2.Points() {
+				closestPt, inside := ClosestTriangleInsidePoint(tri1, pt)
+				if inside && closestPt.Sub(pt).Norm() <= collisionBufferMM {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+// distanceFromMesh returns the minimum distance between this mesh and another mesh.
+func (m *Mesh) distanceFromMesh(other *Mesh) float64 {
+	minDist := math.Inf(1)
+	for _, tri1 := range m.triangles {
+		for _, tri2 := range other.triangles {
+			for _, pt := range tri1.Points() {
+				closestPt, _ := ClosestTriangleInsidePoint(tri2, pt)
+				dist := closestPt.Sub(pt).Norm()
+				if dist < minDist {
+					minDist = dist
+				}
+			}
+			for _, pt := range tri2.Points() {
+				closestPt, _ := ClosestTriangleInsidePoint(tri1, pt)
+				dist := closestPt.Sub(pt).Norm()
+				if dist < minDist {
+					minDist = dist
+				}
+			}
+		}
+	}
+	return minDist
```

Comment:
note to self: review all functions between this and `distanceFromSphere` when you better understand how triangleInsidePoint works

Reply:
Pause for a moment; just realized what I did here isn't sufficient. I need to check the segments between the vertices, not just the vertices themselves.